### DISCUSSION TITLE
Swift native Data instead of Array<UInt8>

### DISCUSSION THREAD

**Discussion by Craz1k0ek** [krzyzanowskim/CryptoSwift#965] [General]
Swift's `Data` struct is, in my opinion, a better way to represent binary data, such as keys, cipher texts and others. Most common Swift functions use `Data` (think `URLSession`, `String.data(using:)`, `Codable`). I would like to see a transition from the `Array<UInt8>` to the `Data` type when possible. This can be implemented without breaking existing functionality.

**Comment by krzyzanowskim** [comment]
When CryptoSwift was created, Swift didn't have a Foundation. Then I added `Foundation` extensions, but `[UInt8]` is still the most basic data type. What I'm afraid of is that `Data` may have a bad performance impact here, but didn't dig that part. (The access to bytes in `Data` is way less convenient) After all `Foundation.Data` and `Array<UInt8>` are easily convertible.

**Comment by Craz1k0ek** [comment]
Worth investigating, looking at where we're standing right now. How do you feel about adding, potentially heuristic, wrappers to the existing API, for example:
```swift
public init(key: Data, blockMode: BlockMode, padding: Padding = .pkcs7) throws {
    try init(key: Array(key), blockMode: blockMode, padding: padding)
}
```

Is it worth working that into a PR, or is that never a direction you'd like to take CryptoSwift?

**Comment by krzyzanowskim** [comment]
I think Xcode will just go banana with these overloads tbh, but give it a try, and see how it works for your own maybe

**Comment by Enricoza** [comment]
Just curious, but shouldn't swift 5.4 allow that kind of overload now?

**Comment by nathanfallet** [comment]
@Craz1k0ek Having initializers with `Data` implies that you `import Foundation`

**Comment by Craz1k0ek** [comment]
```
.macOS(.v10_12), .iOS(.v9), .tvOS(.v9), .watchOS(.v2)
```
@NathanFallet according to the SPM specification, that shouldn't matter

**Comment by nathanfallet** [comment]
@Craz1k0ek Then what about replacing `Array<UInt8>` by `Data` everywhere and create an extension so that `Data` conforms to `ExpressibleByArrayLiteral` protocol? (So that we can use both without problem)

**Comment by krzyzanowskim** [comment]
> Then what about replacing Array<UInt8> by Data everywhere

no. Data on Apple platforms is backed by ObjC NSData and this is the last thing I want to deal with. There's nothing wrong with [UInt8]. Also, the `Data` interface is harder to use it effectively.

**Comment by Craz1k0ek** [comment]
It appears the owner of CryptoSwift is strongly opinionated about the `Data` type that Apple platforms adopt. I doubt that changes to the `Array<UInt8>` will be made any time soon. I have no time to implement it myself, so for now I'll close it.
